leaderelection/context.go

```diff
@@ -173,29 +175,50 @@ func (b *standardBuilder) buildElector(ctx context.Context, la reconciler.Leader
 
 type statefulSetBuilder struct {
 	lec ComponentConfig
-	ssc statefulSetConfig
+	bkt reconciler.Bucket
 }
 
 func (b *statefulSetBuilder) buildElector(ctx context.Context, la reconciler.LeaderAware, enq func(reconciler.Bucket, types.NamespacedName)) (Elector, error) {
 	logger := logging.FromContext(ctx)
-	ordinal := uint32(b.ssc.StatefulSetID.ordinal)
-	logger.Infof("%s will run in StatefulSet ordinal assignement mode with ordinal %d",
-		b.lec.Component, ordinal)
+	logger.Infof("%s will run in StatefulSet ordinal assignement mode with bucket name %s",
+		b.lec.Component, b.bkt.Name())
 
 	return &unopposedElector{
-		bkt: &bucket{
-			// The name is the full pod DNS of the owner pod of this bucket.
-			name: fmt.Sprintf("%s://%s-%d.%s.%s.svc.%s:%s", b.ssc.Protocol,
-				b.ssc.StatefulSetID.ssName, ordinal, b.ssc.ServiceName,
-				system.Namespace(), network.GetClusterDomainName(), b.ssc.Port),
-			index: ordinal,
-			total: b.lec.Buckets,
-		},
+		bkt: b.bkt,
 		la:  la,
 		enq: enq,
 	}, nil
 }
 
+// NewStatefulSetBucketAndSet creates a BucketSet for StatefulSet controller with
+// the given bucket size and the information from environment variables. Then uses
+// the created BucketSet to create a Bucket for this StatefulSet Pod.
+func NewStatefulSetBucketAndSet(buckets int) (reconciler.Bucket, *hash.BucketSet, error) {
+	ssc, err := newStatefulSetConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if ssc.StatefulSetID.ordinal >= buckets {
+		return nil, nil, fmt.Errorf("ordinal %d is out of range [0, %d)",
+			ssc.StatefulSetID.ordinal, buckets)
+	}
+
+	names := sets.String{}
```

Comment:
Does this support `make(sets.String, buckets)` since it's just a type alias of map? 🤔

Reply:
yes. Change.